src/libmeasurement_kit/ooni/telegram.cpp

```diff
@@ -0,0 +1,194 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "../common/utils.hpp"
+#include "../ooni/constants.hpp"
+#include <measurement_kit/ooni.hpp>
+
+namespace mk {
+namespace ooni {
+
+using namespace mk::report;
+
+static void tcp_many(const std::vector<std::string> ip_ports,
+                     Var<Entry> entry,
+                     Callback<Error> cb,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger) {
+    // C++20 should have continuations in std; until then,
+    // we copy the ghetto pattern from web_connectivity
+    int ips_count = ip_ports.size();
+    Var<int> ips_tested(new int(0));
+    if (ips_count == *ips_tested) {
+        cb(NoError());
+        return;
+    }
+
+    // per-endpoint, we only consider success/failure.
+    // telegram is "blocked" if no endpoints succeed.
+    auto tcp_cb = [=](std::string ip, int port) {
+        return [=](Error err, Var<net::Transport> txp) {
+            bool close_txp = true; // if connected, we must disconnect
+            Entry result = {
+                {"ip", ip},
+                {"port", port},
+                {"status",
+                    {{"success", nullptr},
+                    {"failure", nullptr}}},
+            };
+            if (!!err) {
+                logger->info("telegram: failure TCP connecting to %s:%d",
+                             ip.c_str(), port);
+                result["status"]["success"] = false;
+                result["status"]["failure"] = err.as_ooni_error();
+                close_txp = false;
+            } else {
+                logger->info("telegram: success TCP connecting to %s:%d",
+                             ip.c_str(), port);
+                result["status"]["success"] = true;
+            }
+            (*entry)["tcp_connect"].push_back(result);
+            *ips_tested += 1;
```

Comment:
Is this racy? See, for example, https://code.facebook.com/posts/1661982097368498 `multiFooAsync()`. I guess it's not if we're single-threaded and there are no exceptions or IO operations in this code?